### ISSUE TITLE
🔮 Implement Coalition Prediction Model Based on Decision Flow and Voting Patterns

### ISSUE THREAD

**Issue by pethers** [Hack23/cia#7909]
## 🎯 Objective
Implement coalition prediction model based on historical decision flow data, voting patterns, and party alignment matrices to forecast government formation and coalition stability using machine learning techniques.

## 📋 Background
**Current State:**
- ParliamentDecisionFlowPageModContentFactoryImpl shows historical decision flows
- No predictive intelligence for coalition formation
- Missing early warning system for coalition instability
- Cannot forecast voting outcomes based on historical patterns
- No automated detection of emerging party alignments

**Gap Analysis per DATA_ANALYSIS_INTOP_OSINT.md:**
- **Predictive Intelligence Gap**: Framework described but not implemented for coalitions
- **Strategic Intelligence Gap**: Missing government formation forecasting
- **Network Analysis Gap**: Party relationship mapping not visualized
- **Business Value**: High demand from political consultants (€15M market segment)

**Business Impact:**
- 🔮 **Predictive Intelligence**: Forecast coalition formation 3-6 months ahead
- 💼 **Consulting Value**: Premium feature for political consulting segment
- 📊 **Strategic Planning**: Early warnings for coalition instability
- 🎯 **Election Insights**: Predict post-election government scenarios

## 📊 Available Data for Modeling
- **Historical Voting Data**: 10+ years of party voting patterns
- **Decision Flow Data**: Committee→Decision pathway success rates
- **Party Alignment Scores**: Calculated from view_riksdagen_party_ballot_support_annual_summary
- **Coalition History**: Previous government formations (2010-present)
- **Election Results**: Party seat counts and vote percentages

## ✅ Acceptance Criteria
- [ ] Implement coalition prediction ML model using historical voting data
- [ ] Add "Coalition Prediction" view to ParliamentDecisionFlowPageModContentFactoryImpl
- [ ] Visualize party alignment heatmap showing compatibility scores
- [ ] Calculate coalition stability index (0-100 score)
- [ ] Forecast most likely coalition scenarios post-election
- [ ] Include confidence intervals for predictions (80%, 90%, 95%)
- [ ] Add temporal analysis showing alignment changes over time
- [ ] Display feature importance (which factors drive predictions)
- [ ] Export coalition prediction report (PDF) for consultants
- [ ] Model accuracy >75% on historical validation data

## 🛠️ Implementation Guidance

**Files to Create/Modify:**
- `service.impl/src/main/java/com/hack23/cia/service/impl/intelligence/CoalitionPredictionService.java` - ML model service
- `citizen-intelligence-agency/src/main/java/com/hack23/cia/web/impl/ui/application/views/user/parliament/pagemode/CoalitionPredictionPageModContentFactoryImpl.java` - New view
- Add Python ML model: `model/coalition_predictor.py` (call via Jython or REST API)
- Create view: `view_party_coalition_compatibility` for alignment data

**Approach:**

### Phase 1: Feature Engineering (2-3h)
```java
@Service
public class CoalitionFeatureEngineering {
    
    public CoalitionFeatures extractFeatures(String year) {
        // Feature 1: Party voting alignment matrix
        Map<String, Map<String, Double>> alignmentMatrix = calculateVotingAlignment(year);
        
        // Feature 2: Historical coalition participation
        Map<String, Integer> coalitionHistory = getCoalitionParticipation();
        
        // Feature 3: Ideological distance (left-right scale)
        Map<String, Double> ideologyScores = calculateIdeologyScores(year);
        
        // Feature 4: Parliamentary size (seats)
        Map<String, Integer> seatCounts = getParliamentarySeats(year);
        
        // Feature 5: Cross-party committee collaboration
        Map<String, Map<String, Integer>> committeeCollaboration = 
            getCommitteeCollaborationMatrix(year);
        
        return new CoalitionFeatures(
            alignmentMatrix,
            coalitionHistory,
            ideologyScores,
            seatCounts,
            committeeCollaboration);
    }
    
    private Map<String, Map<String, Double>> calculateVotingAlignment(String year) {
        // Query voting patterns from database
        List<PartyVotingRecord> votingRecords = votingRepository
            .findByYear(year);
        
        Map<String, Map<String, Double>> alignmentMatrix = new HashMap<>();
        List<String> parties = List.of("S", "M", "SD", "C", "V", "KD", "MP", "L");
        
        for (String party1 : parties) {
            alignmentMatrix.put(party1, new HashMap<>());
            for (String party2 : parties) {
                if (party1.equals(party2)) {
                    alignmentMatrix.get(party1).put(party2, 1.0);
                    continue;
                }
                
                // Calculate alignment: % of ballots where parties voted same way
                double alignment = calculatePairwiseAlignment(
                    votingRecords, party1, party2);
                
                alignmentMatrix.get(party1).put(party2, alignment);
            }
        }
        
        return alignmentMatrix;
    }
}
```

### Phase 2: ML Model Implementation (3-4h)
```python
# model/coalition_predictor.py
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

class CoalitionPredictor:
    def __init__(self):
        self.model = RandomForestClassifier(
            n_estimators=100,
            max_depth=10,
            random_state=42)
        self.feature_names = []
    
    def train(self, historical_data):
        """
        Train coalition prediction model on historical data
        
        Features:
        - Party alignment matrix (S-M, S-SD, M-SD, etc.) - 28 features
        - Parliamentary seats per party - 8 features
        - Ideological distance - 8 features
        - Historical coalition participation - 8 features
        - Committee collaboration scores - 28 features
        
        Target: Coalition formed (1) or not (0) for each party pair
        """
        X = historical_data[self.feature_names]
        y = historical_data['coalition_formed']
        
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42)
        
        self.model.fit(X_train, y_train)
        
        y_pred = self.model.predict(X_test)
        accuracy = accuracy_score(y_test, y_pred)
        
        print(f"Model Accuracy: {accuracy:.2%}")
        print("\nFeature Importance:")
        for feature, importance in sorted(
                zip(self.feature_names, self.model.feature_importances_),
                key=lambda x: x[1],
                reverse=True)[:10]:
            print(f"  {feature}: {importance:.4f}")
        
        return accuracy
    
    def predict_coalition_scenarios(self, current_features, election_results):
        """
        Predict most likely coalition scenarios
        """
        scenarios = []
        parties = ['S', 'M', 'SD', 'C', 'V', 'KD', 'MP', 'L']
        
        # Generate all possible coalition combinations (2-4 parties)
        from itertools import combinations
        for r in range(2, 5):
            for coalition in combinations(parties, r):
                # Check if coalition has majority (175+ seats)
                total_seats = sum(election_results.get(p, 0) for p in coalition)
                if total_seats < 175:
                    continue
                
                # Extract features for this coalition
                features = self._extract_coalition_features(
                    coalition, current_features)
                
                # Predict formation probability
                probability = self.model.predict_proba([features])[0][1]
                
                # Calculate stability index based on alignment scores
                stability_index = self._calculate_stability(
                    coalition, current_features)
                
                scenarios.append({
                    'coalition': list(coalition),
                    'seats': total_seats,
                    'probability': probability,
                    'stability_index': stability_index
                })
        
        # Sort by probability descending
        scenarios.sort(key=lambda x: x['probability'], reverse=True)
        
        return scenarios[:5]  # Top 5 most likely scenarios
    
    def _calculate_stability(self, coalition, features):
        """
        Calculate coalition stability index (0-100)
        Based on:
        - Voting alignment within coalition
        - Ideological distance
        - Historical cooperation
        """
        alignment_scores = []
        for i, party1 in enumerate(coalition):
            for party2 in coalition[i+1:]:
                alignment = features['alignment'][party1][party2]
                alignment_scores.append(alignment)
        
        avg_alignment = np.mean(alignment_scores)
        
        # Stability = weighted average of alignment factors
        stability = (
            avg_alignment * 0.6 +  # 60% voting alignment
            (1 - self._ideological_variance(coalition, features)) * 0.3 +  # 30% ideology
            self._historical_cooperation(coalition, features) * 0.1  # 10% history
        )
        
        return int(stability * 100)
```

### Phase 3: Coalition Prediction View (2-3h)
```java
@Component
public final class CoalitionPredictionPageModContentFactoryImpl 
        extends AbstractParliamentPageModContentFactoryImpl {
    
    @Autowired
    private CoalitionPredictionService coalitionPredictionService;
    
    @Secured({ "ROLE_USER", "ROLE_ADMIN" })
    @Override
    public Layout createContent(final String parameters, final MenuBar menuBar, final Panel panel) {
        final VerticalLayout panelContent = createPanelContent();
        
        setupMenuAndHeader(menuBar, panel, panelContent);
        
        // Year selector
        final String selectedYear = extractSelectedYear(parameters);
        addYearSelector(panelContent, selectedYear);
        
        // Coalition prediction scenarios
        addCoalitionScenarios(panelContent, selectedYear);
        
        // Party alignment heatmap
        addPartyAlignmentHeatmap(panelContent, selectedYear);
        
        // Feature importance visualization
        addFeatureImportance(panelContent);
        
        return panelContent;
    }
    
    private void addCoalitionScenarios(final VerticalLayout panelContent, final String year) {
        final List<CoalitionScenario> scenarios = coalitionPredictionService
            .predictCoalitions(year);
        
        final VerticalLayout scenariosLayout = new VerticalLayout();
        scenariosLayout.setCaption("🔮 Predicted Coalition Scenarios");
        
        int rank = 1;
        for (CoalitionScenario scenario : scenarios) {
            final HorizontalLayout scenarioRow = new HorizontalLayout();
            scenarioRow.setWidth("100%");
            scenarioRow.setSpacing(true);
            
            // Rank badge
            final Label rankLabel = new Label("#" + rank++);
            rankLabel.addStyleName(ValoTheme.LABEL_H3);
            rankLabel.setWidth("50px");
            
            // Coalition parties
            final Label partiesLabel = new Label(
                String.join(" + ", scenario.getParties()));
            partiesLabel.addStyleName(ValoTheme.LABEL_BOLD);
            partiesLabel.setWidth("200px");
            
            // Probability bar
            final ProgressBar probabilityBar = new ProgressBar();
            probabilityBar.setValue(scenario.getProbability());
            probabilityBar.setCaption(
                String.format("%.1f%% probability", scenario.getProbability() * 100));
            probabilityBar.setWidth("200px");
            
            // Stability index
            final Label stabilityLabel = new Label(
                String.format("Stability: %d/100", scenario.getStabilityIndex()));
            stabilityLabel.addStyleName(
                scenario.getStabilityIndex() > 70 ? ValoTheme.LABEL_SUCCESS : ValoTheme.LABEL_WARNING);
            
            // Seats
            final Label seatsLabel = new Label(
                String.format("%d seats", scenario.getTotalSeats()));
            
            scenarioRow.addComponents(rankLabel, partiesLabel, probabilityBar, stabilityLabel, seatsLabel);
            scenariosLayout.addComponent(scenarioRow);
        }
        
        panelContent.addComponent(scenariosLayout);
    }
    
    private void addPartyAlignmentHeatmap(final VerticalLayout panelContent, final String year) {
        // Create heatmap showing party-party alignment scores
        final Chart heatmap = new Chart(ChartType.HEATMAP);
        heatmap.setHeight("500px");
        
        final Configuration config = heatmap.getConfiguration();
        config.setTitle("Party Alignment Matrix (Voting Patterns " + year + ")");
        
        final List<String> parties = List.of("S", "M", "SD", "C", "V", "KD", "MP", "L");
        config.getxAxis().setCategories(parties.toArray(new String[0]));
        config.getyAxis().setCategories(parties.toArray(new String[0]));
        
        final DataSeries series = new DataSeries();
        final Map<String, Map<String, Double>> alignmentMatrix = 
            coalitionPredictionService.getAlignmentMatrix(year);
        
        for (int i = 0; i < parties.size(); i++) {
            for (int j = 0; j < parties.size(); j++) {
                String party1 = parties.get(i);
                String party2 = parties.get(j);
                double alignment = alignmentMatrix.get(party1).get(party2);
                
                DataSeriesItem item = new DataSeriesItem(i, j, alignment);
                item.setColor(getColorForAlignment(alignment));
                series.add(item);
            }
        }
        
        config.addSeries(series);
        panelContent.addComponent(heatmap);
    }
}
```

### Phase 4: Model Training & Validation (1-2h)
```java
@Service
public class CoalitionModelTrainingService {
    
    @PostConstruct
    public void trainModel() {
        // Fetch historical data (2010-2024)
        List<HistoricalCoalitionData> historicalData = 
            loadHistoricalCoalitionData(2010, 2024);
        
        // Convert to training format
        DataFrame trainingData = prepareTrainingData(historicalData);
        
        // Train model via Jython or REST API to Python service
        CoalitionPredictor predictor = new CoalitionPredictor();
        double accuracy = predictor.train(trainingData);
        
        if (accuracy < 0.75) {
            throw new ModelTrainingException(
                "Model accuracy too low: " + accuracy);
        }
        
        // Save trained model
        saveModel(predictor);
        
        log.info("Coalition prediction model trained successfully. Accuracy: {}%", accuracy * 100);
    }
}
```

## 📊 Expected Intelligence Improvements
| Metric | Current | Target | Improvement |
|--------|---------|--------|-------------|
| Coalition Forecasting | None | 3-6 months ahead | Predictive capability |
| Prediction Accuracy | N/A | >75% | Historical validation |
| Stability Assessment | Manual | Automated (0-100 index) | Quantified risk |
| Scenario Analysis | None | Top 5 likely scenarios | Strategic planning |
| Feature Transparency | N/A | Feature importance shown | Explainable AI |
| Consulting Value | Basic data | Premium ML insights | €15M market segment |

## 🤖 Recommended Agent
**Agent**: @hack23-intelligence-operative + @hack23-stack-specialist
**Rationale**: Requires political intelligence expertise (coalition dynamics) and ML implementation skills (Python/Java integration).

For implementation:
- Intelligence Operative: Design coalition features, validate predictions with domain knowledge
- Stack Specialist: Implement ML model, integrate Python with Java backend, optimize queries

## 🔗 Related Resources
- [DATA_ANALYSIS_INTOP_OSINT.md - Predictive Intelligence](https://github.com/Hack23/cia/blob/master/DATA_ANALYSIS_INTOP_OSINT.md)
- [BUSINESS_PRODUCT_DOCUMENT.md - Political Consulting Market](https://github.com/Hack23/cia/blob/master/BUSINESS_PRODUCT_DOCUMENT.md)
- [ParliamentDecisionFlowPageModContentFactoryImpl.java](https://github.com/Hack23/cia/blob/master/citizen-intelligence-agency/src/main/java/com/hack23/cia/web/impl/ui/application/views/user/parliament/pagemode/ParliamentDecisionFlowPageModContentFactoryImpl.java)
- [Scikit-learn Random Forest](https://scikit-learn.org/stable/modules/ensemble.html#forest)

## 📊 Metadata
**Priority:** Medium | **Effort:** L (12-16h) | **Impact:** Predictive intelligence, premium consulting feature, €15M market opportunity

**Labels:** `type:feature`, `domain:intelligence`, `domain:backend`, `priority:medium`, `size:large`
